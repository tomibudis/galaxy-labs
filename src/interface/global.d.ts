declare module '*.module.scss' {
  interface ClassName {
    [className: string]: string;
  }
  const classNames: ClassName;
  export = classNames;
}
