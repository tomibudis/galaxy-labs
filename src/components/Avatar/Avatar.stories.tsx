import React from 'react';

import { Meta, ComponentStory } from '@storybook/react';

import Avatar from './Avatar';

export default {
  title: 'Components/Avatar',
  component: Avatar,
  argsTypes: {
    control: { table: { disable: true } },
  },
} as Meta<typeof Avatar>;

const Template: ComponentStory<typeof Avatar> = (args): JSX.Element => {
  return <Avatar {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  name: 'Tomi Budi',
  size: 'sm',
};

export const Rectangle = Template.bind({});
Rectangle.args = {
  name: 'John Doe',
  type: 'rectangle',
  icon: null,
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  icon: 'person',
  className: 'bg-warning',
};

export const WithImage = Template.bind({});
WithImage.args = {
  img: 'https://www.signivis.com/img/custom/avatars/member-avatar-01.png',
};

export const Size = (): JSX.Element => {
  return (
    <div className="d-flex gap-3">
      <Avatar size="sm" name="Small" />
      <Avatar size="md" name="Medium" />
      <Avatar size="lg" name="Large" />
    </div>
  );
};
