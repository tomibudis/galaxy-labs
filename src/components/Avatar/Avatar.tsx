import React, { useCallback, useMemo } from 'react';

import PropTypes from 'prop-types';
import cx from 'classnames';

import { AvatarProps } from '#interface/index';
import { TYPE_AVATAR, SIZE } from '#constants/index';
import Icon from '../Icon';

import styles from './styles.module.scss';

const Avatar: React.FC<AvatarProps> = (props): JSX.Element => {
  const getAbrebativeName = useCallback(() => {
    if (!props.name) return null;
    const maxChar = 2;
    const sliceName = props.name
      .split(' ')
      .map((splitName, index) => {
        if (index + 1 > maxChar) return null;
        return splitName.slice(0, 1);
      })
      .join('');
    return sliceName;
  }, [props.name]);

  const renderIcon = useCallback(() => {
    if (!props.icon) return null;
    return <Icon name={props.icon} />;
  }, [props.icon]);

  const stylesImage = useMemo(() => {
    if (!props.img) return {};
    return { backgroundImage: `url(${props.img})` };
  }, [props.img]);

  const renderName = useMemo(() => {
    if (props.img) return null;
    return props.name ? getAbrebativeName() : renderIcon();
  }, [props.img, props.name, getAbrebativeName, renderIcon]);

  return (
    <div
      className={cx(
        { [styles.avatarRectangle]: props.type === TYPE_AVATAR.RECTANGLE },
        { [styles.avatarCircle]: props.type === TYPE_AVATAR.CIRCLE },
        { [styles.sm]: props.size === SIZE.SM },
        { [styles.md]: props.size === SIZE.MD },
        { [styles.lg]: props.size === SIZE.LG },
        props.className
      )}
      style={stylesImage}
    >
      {renderName}
    </div>
  );
};

Avatar.propTypes = {
  /** Set size of avatar with 3 variant */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  /** Set name of avatar */
  name: PropTypes.string,
  /** Set name of icon if we want to using icon instead of alphabetics */
  icon: PropTypes.string,
  /** Set type of box avatar */
  type: PropTypes.oneOf(['rectangle', 'circle']),
};

Avatar.defaultProps = {
  type: 'circle',
  size: 'sm',
};

export default Avatar;
