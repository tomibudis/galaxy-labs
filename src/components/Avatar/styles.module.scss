.choreStyles {
  background: #ccc;
  color: #fff;
  width: 35px;
  height: 35px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
}

.avatarRectangle {
  /* stylelint-disable-next-line at-rule-no-unknown */
  @extend .choreStyles;

  border-radius: 6px;
}

.avatarCircle {
  /* stylelint-disable-next-line at-rule-no-unknown */
  @extend .choreStyles;

  border-radius: 50%;
}

.sm {
  width: 35px;
  height: 35px;
}

.md {
  width: 50px;
  height: 50px;
  font-size: large;
}

.lg {
  width: 70px;
  height: 70px;
  font-size: x-large;
}
