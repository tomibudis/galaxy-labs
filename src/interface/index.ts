import React from 'react';

import * as moment from 'moment';
import { Options } from 'react-select';

type SizeTypes = 'sm' | 'md' | 'lg';

type VariantTypes = 'primary' | 'secondary' | 'warning' | 'danger';

type Level = 1 | 2 | 3 | 4 | 5 | 6;

type Option = {
  label: string;
  value: string;
};

export interface ButtonProps {
  isLoading?: boolean;
  variant?: VariantTypes;
  size?: SizeTypes;
  children: React.ReactNode;
}

export interface HeadingProps {
  level?: Level;
  className?: string;
  children: React.ReactNode | string;
  [x: string]: unknown;
}

export interface IconProps {
  name: string;
  className?: string;
  size?: SizeTypes;
}

export interface SelectProps {
  name: string;
  isLoading?: boolean;
  isClearable?: boolean;
  isSearchable?: boolean;
  className?: string;
  classNamePrefix?: string;
  defaultValue?: string;
  isDisabled?: boolean;
  options: Options<Option>;
}

export interface ModalProps {
  isVisible: boolean | undefined;
  onClose: () => void;
  size: SizeTypes;
  className: string;
}

export interface TextProps {
  className?: string;
  children: React.ReactNode;
}

export interface LabelProps {
  className?: string;
  children: React.ReactNode;
  id?: string | undefined;
  htmlFor?: string | undefined;
  isRequired?: boolean | undefined;
}

export interface CalendarProps {
  date?: moment.Moment | null | undefined;
  noBorder?: boolean | undefined;
  hideKeyboardShortcutsPanel?: boolean | undefined;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  onDateChange?: (selectedDate: moment.Moment) => moment.Moment;
  onOutsideClick?: () => React.MouseEvent<HTMLElement>;
  onPrevMonthClick?: () => moment.Moment;
  onNextMonthClick?: () => moment.Moment;
  locale?: string;
}

export interface AvatarProps {
  size?: SizeTypes;
  name?: string;
  icon?: string;
  type: 'rectangle' | 'circle' | undefined;
  className?: string;
  img?: string;
}
