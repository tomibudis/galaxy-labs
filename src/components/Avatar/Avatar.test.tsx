import React from 'react';

import { cleanup, render } from '@testing-library/react';

import { AvatarProps } from '../../interface/index';
import Avatar from './Avatar';

afterEach(() => {
  cleanup();
});

const AvatarComponent: React.FC<AvatarProps> = (props): JSX.Element => {
  return <Avatar {...props} />;
};

describe('<Avatar />', () => {
  it('render component without crashing', () => {
    render(<AvatarComponent type="circle" />);
    expect(true).toBeTruthy();
  });

  it('render type rectangle avatar correctly', () => {
    const { container } = render(<AvatarComponent type="rectangle" />);
    expect(container.firstChild).toHaveClass('avatarRectangle');
  });

  it('render type circle avatar correctly', () => {
    const { container } = render(<AvatarComponent type="circle" />);
    expect(container.firstChild).toHaveClass('avatarCircle');
  });

  it('match with snapshot file', () => {
    const { baseElement } = render(
      <AvatarComponent type="circle" name="Tomi Budi" />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
