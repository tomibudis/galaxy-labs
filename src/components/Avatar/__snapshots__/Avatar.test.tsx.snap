// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> match with snapshot file 1`] = `
<body>
  <div>
    <div
      class="avatarCircle sm"
    >
      TB
    </div>
  </div>
</body>
`;
